### PR TITLE
Windows builds are broken due to MSVC fold expression bug

### PR DESCRIPTION
**Pull request by nolankramer**
This is addressed by adding workaround type traits into `TypeTraits.hpp`:

```
template<typename T, typename... Ts>
struct all_are_node_ptrs {
    static constexpr bool value = (is_node_ptr_v<T> && ... && is_node_ptr_v<Ts>);
};

template<typename T, typename... Ts>
inline constexpr bool all_are_node_ptrs_v = all_are_node_ptrs<T, Ts...>::value;
```

and

```
template<typename T, typename... Ts>
struct all_are_edge_ptrs {
    static constexpr bool value = (is_edge_ptr_v<T> && ... && is_edge_ptr_v<Ts>);
};

template<typename T, typename... Ts>
inline constexpr bool all_are_edge_ptrs_v = all_are_edge_ptrs<T, Ts...>::value;
```

I've also cleaned-up warnings on Windows that were caused by not explicitly typecasting.

### CODE REVIEW

**Comment by sbaldu** [comment]
The clang-format workflow fails in files `test/DijkstraTest.cpp`, `TypeTraitsTest.hpp` and `welshPowellColoringTest.cpp`

**Comment by nolankramer** [comment]
@sbaldu Ran it. Expect a high number of changes! Seems quite a few files weren't formatted correctly.

**Comment by nolankramer** [comment]
Looks like format is still broken on `include/`???

I wonder why. I'm using a freshly installed `clang-format` on WSL Ubuntu: `Ubuntu clang-format version 14.0.0-1ubuntu1.1`

**Comment by sbaldu** [comment]
The workflow uses `clang-format 16`, maybe it's due to some difference in the two versions. It's happened to me before when working with `clang-format 14`

**Comment by nolankramer** [comment]
Hmm let me try to get that version

**Comment by sbaldu** [comment]
@nolankramer  I see that partition test is failing on windows, but I suppose that the issue is unrelated to this PR.

**Comment by nolankramer** [comment]
> @nolankramer I see that partition test is failing on windows, but I suppose that the issue is unrelated to this PR.

FWIW, that test has been failing for a long time. Somebody is grabbing a mutex they don't own in the partitioning code.

**Comment by sbaldu** [comment]
> > @nolankramer I see that partition test is failing on windows, but I suppose that the issue is unrelated to this PR.
> 
> FWIW, that test has been failing for a long time. Somebody is grabbing a mutex they don't own in the partitioning code. That can likely be cleaned up in another PR.

Yes you're right.

**Comment by sbaldu** [comment]
@nolankramer Perfect

**Comment by nolankramer** [comment]
This can be merged after checks. Looks like the failing ones are due to infra bugs.

**Comment by nolankramer** [comment]
@sbaldu This can be merged now.

**Comment by sbaldu** [comment]
I agree. I'm waiting for @ZigRazor approval.

**Review by ZigRazor — APPROVED** [review]
Good one!

**Comment by ZigRazor** [comment]
I think we can open a new issue for partitioning test failed, that you cite before @nolankramer

**Pull request merged**
